README.md
### REQUIREMENTS:

- [.NET Framework 4.7.2 Runtime](https://dotnet.microsoft.com/download/dotnet-framework/net472)
- [Microsoft Visual C++ 2015-2019 Re-distributable](https://aka.ms/vs/16/release/vc_redist.x64.exe)

---

### INSTALLATION:

1. build it
2. install it
3. enjoy lemons

### MANUAL UNINSTALL:

1. Make sure the Game you are installing LemonLoader to is Closed and Not Running.
2. Remove the version.dll file from the Game's Installation Folder.
3. Remove the LemonLoader folder from the Game's Installation Folder.
4. Remove the Plugins folder from the Game's Installation Folder.
5. Remove the Mods folder from the Game's Installation Folder.
6. Remove the Logs folder from the Game's Installation Folder.

---

### GENERAL INFORMATION:

- All Plugins get placed in the created Plugins folder inside the Game's Install Folder.
- All Mods get placed in the created Mods folder inside the Game's Install Folder.
- All Logs are made in the created Logs folder inside the Game's Install Folder.

---

### CONSOLE DIFFERENCES:

- Normal Console is for Debugging and Displaying the Output of Plugins and Mods.
- Debug Mode/Console is for Debugging LemonLoader Internals.

---

### LAUNCH OPTIONS:

| Argument              | Description                              |
| --------------------- | ---------------------------------------- |
| --no-mods | Launch game without loading Mods |
| --quitfix | Fixes the Hanging Process Issue with some Games |
| --LemonLoader.hideconsole | Hides the Normal Console |
| --LemonLoader.hidewarnings | Hides Warnings from Displaying |
| --LemonLoader.debug | Debug Mode/Console |
| --LemonLoader.magenta | Magenta Console Color |
| --LemonLoader.rainbow | Rainbow Console Color |
| --LemonLoader.randomrainbow | Random Rainbow Console Color |
| --LemonLoader.maxlogs | Max Log Files  [ Default: 10 ] [ Disable: 0 ] |
| --LemonLoader.maxwarnings | Max Warnings per Log File  [ Default: 100 ] [ Disable: 0 ] |
| --LemonLoader.maxerrors | Max Errors per Log File  [ Default: 100 ] [ Disable: 0 ] |
| --LemonLoader.devpluginsonly | Loads only Plugins with the "-dev.dll" extension |
| --LemonLoader.devmodsonly | Loads only Mods with the "-dev.dll" extension |
| --LemonLoader.agregenerate | Forces Assembly to be Regenerated on Il2Cpp Games |
| --LemonLoader.agfvunhollower | Forces the Assembly Generator to use a Specified Version of Il2CppAssemblyUnhollower |

---

### CREDITS+LICENSING:

LemonLoader is licensed under the Apache License, Version 2.0. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LICENSE.md) for the full License.  
[LemonLoader.dll](LemonLoader) was adapted from [AtiRoNya](https://github.com/AtiLion/AtiRoNya) by [AtiLion](https://github.com/AtiLion), licensed under [MIT License](https://github.com/AtiLion/AtiRoNya/blob/e20e4a8fc47b37834c8284f9e6e937f04a84c510/LICENSE)

Third-party libraries used as source code or bundled in binary form:
- [Research Detours Package](https://github.com/microsoft/Detours) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/Detours/LICENSE.md) for the full License.
- [Mono](https://github.com/Unity-Technologies/mono) is licensed under multiple licenses. See [LICENSE](https://github.com/Unity-Technologies/mono/blob/unity-master/LICENSE) for full details.
- [Harmony](https://github.com/pardeike/Harmony) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LemonLoader.ModHandler/Harmony/LICENSE) for the full License.
- [TinyJSON](https://github.com/pbhogan/TinyJSON) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LemonLoader.AssemblyGenerator/TinyJSON/LICENSE.md) for the full License.
- [LightJson](https://github.com/MarcosLopezC/LightJson) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LemonLoader.Installer/LightJson/LICENSE.txt) for the full License.
- [Tomlyn](https://github.com/xoofx/Tomlyn) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LemonLoader.ModHandler/Tomlyn/license.txt) for the full License.
- [SharpZipLib](https://github.com/icsharpcode/SharpZipLib) is licensed under the MIT License. See [LICENSE](https://github.com/HerpDerpinstine/LemonLoader/blob/master/LemonLoader.ModHandler/SharpZipLib/LICENSE.txt) for the full License.
- Unity Runtime libraries (found in [Unity Dependencies](BaseLibs/Unity%20Dependencies)) are part of Unity Software.  
Their usage is subject to [Unity Terms of Service](https://unity3d.com/legal/terms-of-service), including [Unity Software Additional Terms](https://unity3d.com/legal/terms-of-service/software).

External tools downloaded and used at runtime:
- [Il2CppDumper](https://github.com/Perfare/Il2CppDumper) is licensed under the MIT License. See [LICENSE](https://github.com/Perfare/Il2CppDumper/blob/master/LICENSE) for the full License.
- [Il2CppAssemblyUnhollower](https://github.com/knah/Il2CppAssemblyUnhollower) is licensed under the GNU Lesser General Public License v3.0. See [LICENSE](https://github.com/knah/Il2CppAssemblyUnhollower/blob/master/LICENSE) for the full License.

---

See [LemonLoader Wiki](https://melonwiki.xyz/#/credits) for the full Credits.

LemonLoader is not sponsored by, affiliated with or endorsed by Unity Technologies or its affiliates.  
"Unity" is a trademark or a registered trademark of Unity Technologies or its affiliates in the U.S. and elsewhere.


CHANGELOG.md
### Version List:

[v0.2.3 Open-Beta](#v023-open-beta)  
[v0.2.2 Open-Beta](#v022-open-beta)  
[v0.2.1 Open-Beta](#v021-open-beta)  
[v0.2.0 Open-Beta](#v020-open-beta)  
[v0.1.0 Open-Beta](#v010-open-beta)  
[v0.0.3 Closed-Beta](#v003-closed-beta)  
[v0.0.2 Closed-Beta](#v002-closed-beta)  
[v0.0.1 Closed-Beta](#v001-closed-beta)  

---

### v0.2.3 Open-Beta:

1. Added Unity 2019.3.4 Dependencies.
2. Temporarily Fixed Issue with TLS Fix on Certain Il2Cpp Unity Games.
3. Added Launch Option "--LemonLoader.agfvunhollower" that Forces the Assembly Generator to use a Specified Version of Il2CppAssemblyUnhollower.
4. Added Location property to the MelonMod class.
5. Revised Plugin System.
6. Fixed Issue causing the Assembly Generator to constantly Regenerate every time a Game Launched.
7. Installer now creates the Plugins folder.
8. Revised MelonMod class to use a new MelonBase.
9. All Commonly used Methods and Properties that both MelonMod and MelonPlugin use have been moved to MelonBase.
10. Plugins now use their own Attributes.
11. Fixed Issue that let Plugins load as Mods when placed in the Mods folder and vice-versa.
12. Added better Legacy Version handling for "--LemonLoader.agfvunhollower".
13. Added Clearer Display Text to State that DLL Status is Game Compatibility.
14. Also call Harmony patches if IL2CPP methods get called using reflection / il2cpp_runtime_invoke.  (Credits to zeobviouslyfakeacc :3)
15. Fixed Issue with Harmony Patches on Virtual Methods.  (Credits to zeobviouslyfakeacc :3)
16. Improved the Harmony error message if the unhollowed target method is not backed by an IL2CPP method.  (Credits to zeobviouslyfakeacc :3)
17. Fixed Issue with Harmony AccessTools not being able to resolve the no-args constructor if a static constructor exists.  (Credits to zeobviouslyfakeacc :3)
18. Added Support for Mono Debuggers.
19. Mono JIT Domain on Il2Cpp Games should now parse Command Line arguments properly.
20. Forced Installer to reinstall the entire LemonLoader folder.
21. Moved all Console and Log Functionality to Native Exports.
22. Updated Il2CppDumper to v6.3.0.
23. Updated Il2CppDumper Config Fix in the Assembly Generator to support Il2CppDumper v6.3.0.
24. Added Temp File Cache to Assembly Generator.
25. Re-enabled Launch Option "--LemonLoader.devpluginsonly".
26. Fixed Issue with Certain Mono Games not having OnApplicationStart called.
27. Added Unity 2018.4.2 Dependencies.
28. Fixed Debug Compilation Issue with Console.
29. Added Launch Option "--LemonLoader.hidewarnings" that will Hide Warnings from Displaying in Normal Console.  (Debug Console ignores this Launch Option.)
30. Fixed Issue with Support Modules not Unloading their Components properly when OnApplicationQuit is called.
31. Fixed Mistake with Error Catching Console Creation.
32. Log Spam Shield will now Disable itself when LemonLoader is using Debug Mode/Console.
33. Log Spam Shield has now been moved to the Native Logger.
34. Added Launch Option "--LemonLoader.maxwarnings" for Setting the Max Amount of Warnings for the Log Spam Shield.  (Debug Console ignores this Launch Option, Default is 100, 0 to Disable.)
35. Added Launch Option "--LemonLoader.maxerrors" for Setting the Max Amount of Errors for the Log Spam Shield.  (Debug Console ignores this Launch Option, Default is 100, 0 to Disable.)
36. Fixed Memory Leak in Log Spam Shield.
37. Fixed Issue with Exports.
38. Moved DLL Status to its own Enum.
39. Added "Compatibility" property to MelonBase.
40. Removed "IsUniversal" property from MelonBase.
41. Fixed Issue with Incompatible Plugins and Mods still being used after Logging.
42. Added Unity 2018.4.23 Dependencies.
43. Fixed Issue with Console.h / Console::IsInitialized().
44. Added Tomlyn to ModHandler for TOML Parsing.
45. Moved Tomlyn to its own Namespace under LemonLoader.Tomlyn.
46. Added Unity 2018.3.14 Dependencies.
47. Added Unity 2019.2.16 Dependencies.
48. Added Tomlyn to Assembly Generator for TOML Parsing.
49. Cleaned up Config Handling in Assembly Generator.
50. Fixed Issue with Corrupt Tomlyn Lib.
51. Added Dependency Graph System for Plugins and Mods.  (Credits to zeobviouslyfakeacc :3)
52. Fixed Issue with Plugins still calling OnPreInitialization even when Incompatible.  (Credits to zeobviouslyfakeacc :3)
53. Added Better Plugin and Mod Cleanup when OnPreInitialization or OnApplicationStart fails.  (Credits to zeobviouslyfakeacc :3)
54. Switched Assembly Generator Config from JSON to TOML.
55. Assembly Generator Config will now be saved to a file named "config.cfg" instead of "config.json".
56. Fixed Issue with Tomlyn in the ModHandler.
57. Added Unity 2020.1.0 Dependencies.
58. Added Temporary Warning for Missing Dependencies.  (Credits to zeobviouslyfakeacc :3)
59. Added MelonOptionalDependencies Attribute.  (Credits to zeobviouslyfakeacc :3)
60. Fixed compilation issues on .NET Framework 3.5.  (Credits to zeobviouslyfakeacc :3)
61. Implemented 'Populate' methods to TinyJSON to de-serialize JSON into an existing object.  (Credits to zeobviouslyfakeacc :3)
62. Reimplemented TinyJSON in AssemblyGenerator to be up to date with the one in the ModHandler.
63. Static Linked Runtime Library to Reduce User Issues with Microsoft Visual C++ 2015-2019 Re-distributable.  (Credits to knah :D)
64. Updated Il2CppAssemblyUnhollower Download to v0.4.8.  (Credits to knah :D)
65. Fixed Issue with the RenderHelper of ModSettingsMenu causing Error when SupportModule Unloads their Component.
66. SupportModules now properly Unloads their Component if no Plugins or Mods are loaded.
67. Temporarily Disabled ModSettingsMenu.
68. Added Launch Option "--LemonLoader.chromium" to turn the Console Color Magenta.  (Thanks Chromium lol)
69. Fixed Issue with Logger methods with Color Parameter being Overridden by RainbowMode or ChromiumMode.
70. Added Unity 2019.3.5 Dependencies.
71. Added Unity 5.6.7 Dependencies.
72. Added Unity 5.6.3 Dependencies.

LemonLoader/LemonLoader.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include <memory>
#include "LemonLoader.h"
#include "Console.h"
#include "Mono.h"
#include "HookManager.h"
#include "Logger.h"
#include "ModHandler.h"
#include <signal.h>
#pragma warning(disable:4996)

HINSTANCE LemonLoader::thisdll = NULL;
int LemonLoader::CommandLineC = NULL;
char* LemonLoader::CommandLineV[64];
bool LemonLoader::IsGameIl2Cpp = false;
bool LemonLoader::DebugMode = false;
bool LemonLoader::QuitFix = false;
bool LemonLoader::DevModsOnly = false;
bool LemonLoader::DevPluginsOnly = false;
bool LemonLoader::AG_Force_Regenerate = false;
char* LemonLoader::ExePath = NULL;;
char* LemonLoader::GamePath = NULL;
char* LemonLoader::DataPath = NULL;
char* LemonLoader::CompanyName = NULL;
char* LemonLoader::ProductName = NULL;
char* LemonLoader::ForceUnhollowerVersion = NULL;

void LemonLoader::Main()
{
	if (CheckOSVersion())
	{
		ParseCommandLine();

		LPSTR filepath = new CHAR[MAX_PATH];
		HMODULE exe_module = GetModuleHandle(NULL);
		GetModuleFileName(exe_module, filepath, MAX_PATH);

		std::string filepathstr = filepath;
		ExePath = new char[filepathstr.size() + 1];
		std::copy(filepathstr.begin(), filepathstr.end(), ExePath);
		ExePath[filepathstr.size()] = '\0';

		filepathstr = filepathstr.substr(0, filepathstr.find_last_of("\\/"));
		GamePath = new char[filepathstr.size() + 1];
		std::copy(filepathstr.begin(), filepathstr.end(), GamePath);
		GamePath[filepathstr.size()] = '\0';

		std::string gameassemblypath = filepathstr + "\\GameAssembly.dll";
		WIN32_FIND_DATA data;
		HANDLE h = FindFirstFile(gameassemblypath.c_str(), &data);
		if (h != INVALID_HANDLE_VALUE)
			IsGameIl2Cpp = true;

		Logger::Initialize(filepathstr);

#ifdef DEBUG
		DebugMode = true;
		Console::Create();
#endif

		std::string pdatapath = filepathstr + "\\*_Data";
		h = FindFirstFile(pdatapath.c_str(), &data);
		if (h != INVALID_HANDLE_VALUE)
		{
			char* nPtr = new char[lstrlen(data.cFileName) + 1];
			for (int i = 0; i < lstrlen(data.cFileName); i++)
				nPtr[i] = char(data.cFileName[i]);
			nPtr[lstrlen(data.cFileName)] = '\0';

			std::string ndatapath = filepathstr + "\\" + std::string(nPtr);
			DataPath = new char[ndatapath.size() + 1];
			std::copy(ndatapath.begin(), ndatapath.end(), DataPath);
			DataPath[ndatapath.size()] = '\0';

			std::string assemblypath = std::string();
			std::string basepath = std::string();
			std::string configpath = std::string();
			if (IsGameIl2Cpp)
			{
				assemblypath = filepathstr + "\\LemonLoader\\Managed";
				basepath = filepathstr + "\\LemonLoader\\Dependencies";
				configpath = ndatapath + "\\il2cpp_data\\etc";
			}
			else
			{
				assemblypath = ndatapath + "\\Managed";
				std::string newbasepath = filepathstr + "\\Mono";
				h = FindFirstFile(newbasepath.c_str(), &data);
				if (h == INVALID_HANDLE_VALUE)
				{
					newbasepath = ndatapath + "\\Mono";
					h = FindFirstFile(newbasepath.c_str(), &data);
				}
				if (h == INVALID_HANDLE_VALUE)
				{
					newbasepath = filepathstr + "\\MonoBleedingEdge";
					h = FindFirstFile(newbasepath.c_str(), &data);
				}
				if (h == INVALID_HANDLE_VALUE)
				{
					newbasepath = ndatapath + "\\MonoBleedingEdge";
					h = FindFirstFile(newbasepath.c_str(), &data);
				}
				if (h != INVALID_HANDLE_VALUE)
				{
					basepath = newbasepath + "\\EmbedRuntime";
					configpath = newbasepath + "\\etc";
				}
			}
			if (!assemblypath.empty())
			{
				Mono::AssemblyPath = new char[assemblypath.size() + 1];
				std::copy(assemblypath.begin(), assemblypath.end(), Mono::AssemblyPath);
				Mono::AssemblyPath[assemblypath.size()] = '\0';
			}
			if (!basepath.empty())
			{
				Mono::BasePath = new char[basepath.size() + 1];
				std::copy(basepath.begin(), basepath.end(), Mono::BasePath);
				Mono::BasePath[basepath.size()] = '\0';
			}
			if (!configpath.empty())
			{
				Mono::ConfigPath = new char[configpath.size() + 1];
				std::copy(configpath.begin(), configpath.end(), Mono::ConfigPath);
				Mono::ConfigPath[configpath.size()] = '\0';
			}

			ReadAppInfo();

			if (IsGameIl2Cpp)
			{
				if (Mono::Load() && Mono::Setup())
					HookManager::LoadLibraryW_Hook();
			}
			else
				HookManager::LoadLibraryW_Hook();
		}
	}
}

void LemonLoader::ParseCommandLine()
{
	char* next = NULL;
	char* curchar = strtok_s(GetCommandLine(), " ", &next);
	while (curchar && (CommandLineC < 63))
	{
		CommandLineV[CommandLineC++] = curchar;
		curchar = strtok_s(0, " ", &next);
	}
	CommandLineV[CommandLineC] = 0;
	if (CommandLineC > 0)
	{
		for (int i = 0; i < CommandLineC; i++)
		{
			const char* command = CommandLineV[i];
			if (command != NULL)
			{
				if (strstr(command, "--quitfix") != NULL)
					QuitFix = true;
				else if (strstr(command, "--LemonLoader.magenta"))
					Console::ChromiumMode = true;
				else if (strstr(command, "--LemonLoader.rainbow") != NULL)
					Console::HordiniMode = true;
				else if (strstr(command, "--LemonLoader.randomrainbow") != NULL)
					Console::HordiniMode_Random = true;
				else if (strstr(command, "--LemonLoader.devmodsonly") != NULL)
					DevModsOnly = true;
				else if (strstr(command, "--LemonLoader.devpluginsonly") != NULL)
					DevPluginsOnly = true;
				else if (strstr(command, "--LemonLoader.agregenerate") != NULL)
					AG_Force_Regenerate = true;
				else if (strstr(command, "--LemonLoader.agfvunhollower"))
					ForceUnhollowerVersion = CommandLineV[i + 1];
#ifndef DEBUG
				else if (strstr(command, "--LemonLoader.maxlogs") != NULL)
					Logger::MaxLogs = GetIntFromConstChar(CommandLineV[i + 1], 10);
				else if (strstr(command, "--LemonLoader.maxwarnings") != NULL)
					Logger::MaxWarnings = GetIntFromConstChar(CommandLineV[i + 1], 10);
				else if (strstr(command, "--LemonLoader.maxerrors") != NULL)
					Logger::MaxErrors = GetIntFromConstChar(CommandLineV[i + 1], 10);
				else if (strstr(command, "--LemonLoader.hideconsole") != NULL)
					Console::Enabled = false;
				else if (strstr(command, "--LemonLoader.hidewarnings") != NULL)
					Console::HideWarnings = false;
				else if (strstr(command, "--LemonLoader.debug") != NULL)
				{
					DebugMode = true;
					Console::Create();
				}
#endif
			}
		}
	}
}

void LemonLoader::ReadAppInfo()
{
	std::ifstream appinfofile((std::string(DataPath) + "\\app.info"));
	std::string line;
	while (std::getline(appinfofile, line, '\n'))
	{
		if (CompanyName == NULL)
		{
			CompanyName = new char[line.size() + 1];
			std::copy(line.begin(), line.end(), CompanyName);
			CompanyName[line.size()] = '\0';
		}
		else
		{
			ProductName = new char[line.size() + 1];
			std::copy(line.begin(), line.end(), ProductName);
			ProductName[line.size()] = '\0';
		}
	}
	appinfofile.close();
}

bool LemonLoader::CheckOSVersion()
{
	OSVERSIONINFO info;
	ZeroMemory(&info, sizeof(OSVERSIONINFO));
	info.dwOSVersionInfoSize = sizeof(OSVERSIONINFO);
	GetVersionEx(&info);
	if ((info.dwMajorVersion < 6) || ((info.dwMajorVersion == 6) && (info.dwMinorVersion < 1)))
	{
		int result = MessageBox(NULL, "You are running on an Unsupported OS.\nWe can not offer support if there are any issues.\nContinue with LemonLoader?", "LemonLoader", MB_ICONWARNING | MB_YESNO);
		if (result == IDYES)
			return true;
		return false;
	}
	return true;
}

void LemonLoader::UNLOAD()
{
	HookManager::UnhookAll();
	Logger::Stop();
	if (LemonLoader::QuitFix)
		LemonLoader::KillProcess();
}

void LemonLoader::KillProcess()
{
	HANDLE hProcess = GetCurrentProcess();
	if (hProcess != NULL)
	{
		TerminateProcess(hProcess, NULL);
		CloseHandle(hProcess);
	}
}

int LemonLoader::CountSubstring(std::string pat, std::string txt)
{
	size_t M = pat.length();
	size_t N = txt.length();
	int res = 0;
	for (int i = 0; i <= N - M; i++)
	{
		int j;
		for (j = 0; j < M; j++)
			if (txt[i + j] != pat[j])
				break;
		if (j == M)
		{
			res++;
			j = 0;
		}
	}
	return res;
}

bool LemonLoader::DirectoryExists(const char* path)
{
	struct stat info;
	if (stat(path, &info) != NULL)
		return false;
	if (info.st_mode & S_IFDIR)
		return true;
	return false;
}

long LemonLoader::GetFileSize(std::string filename)
{
	struct stat stat_buf;
	return ((stat(filename.c_str(), &stat_buf) == 0) ? stat_buf.st_size : -1);
}

int LemonLoader::GetIntFromConstChar(const char* str, int defaultval)
{
	if (str == NULL || *str == '\0')
		return defaultval;
    bool negate = (str[0] == '-');
    if ( *str == '+' || *str == '-' )
        ++str;
	if (*str == '\0')
		return defaultval;
    int result = 0;
    while(*str)
    {
		if (*str >= '0' && *str <= '9')
			result = result * 10 - (*str - '0');
		else
			return defaultval;
        ++str;
    }
    return negate ? result : -result;
}